### ISSUE TITLE
Add option to break after = in assignment

### ISSUE THREAD

**Issue by chrisbouchard** [rust-lang/rustfmt#3514]
I would like to see an option in rustfmt to break after the `=` in an assignment when the right-hand-side is "complex". I admit I don't know what "complex" is &mdash; possibly if the right-hand-side is subject to block/visual formatting. Here are a couple examples:

```rust
let var1 = if condition {
    true_part
} else {
    false_part
};

let var2 = some_value
    .chained()
    .method()
    .calls();
```

To me, it would be nicer to see this as:

```rust
let var1 =
    if condition {
        true_part
    } else {
        false_part
    };

let var2 =
    some_value
        .chained()
        .method()
        .calls();
```

In both cases, it seems more readable to me to keep the block together at the same indent level, especially when the variable name gets long.

```rust
let it_gets_harder_to_see_them_together = if condition {
    true_part
} else {
    false_part
};

let some_value_winds_up_in_right_field = some_value
    .chained()
    .method()
    .calls();
```

Compare

```rust
let it_gets_harder_to_see_them_together =
    if condition {
        true_part
    } else {
        false_part
    };

let some_value_winds_up_in_right_field =
    some_value
        .chained()
        .method()
        .calls();
```

You can also replace long variable names with reasonable variables with type hints.

```rust
let var1: SomeType<Type1, Type2> = if condition {
    true_part
} else {
    false_part
};
```

To be clear, I want to keep block formatting. I just would like the option to have the block start indented on the next line.

PS. I didn't want to include this as a motivating example, because it's hardly rustfmt's fault, but editors like IntelliJ will insert "smart" type hints inline automatically. This pads out the line length in a way that rustfmt can't see, and it's why I don't suggest basing this on line length, but rather on whether the right-hand-side is "complex", block-formatted, or something like that.

**Comment by Proksima** [comment]
Indeed, I was surprised when trying to customize my `rustfmt.toml` that there were close to no customization when it comes to assignments. Then again, looking at how everyone code, I think I am the oddball with my

```
let a = {
    fct()
        .foo()
        .bar()
};
```

**Comment by seanpianka** [comment]
I'm also looking for an option to prevent the following line-break after a long expression assignment (once it reaches past the 80-character line-width rule):
```rust
mod tests {
     #[test]
     fn test1() {
         let principal =
             Principal::new(ID::new_v4(), RoleDiscriminants::Consumer);
     }
}
```
I would rather enforce this style with block indentation:
```rust
mod tests {
     #[test]
     fn test1() {
         let principal = Principal::new(
             ID::new_v4(), 
             RoleDiscriminants::Consumer
         );
     }
}
```
Is it possible to enforce arguments are placed on newlines instead of the entire expression put onto a new line?

**Comment by kang-sw** [comment]
+1 for this. Sometimes, async and lambdas need to be differentiated from simple multi-line allocations

 For example:

```rust
let logic_invoked_immediately/* sometimes-great-IDE-supported-long-inlay-hint-inserted-here */ = if let Some(something) = do_something {
    something
}else {
    fallback
};
let logic_run_later/* sometimes-great-IDE-supported-long-inlay-hint-inserted-here */ = |param_1| {
    do_ops_with(&param_1);
    param_1.do_something_else();
    param_1
};

let this_one_differentiates_from_immediately_invoked_logics =
    |param_1|  {
        do_ops_with(&param_1);
        param_1.do_something_else();
        param_1
    };

let this_might_be_executed_later_either =  
   async move {
      do_some_async_calc().await
   };
```

CLion rust plugin formatter is much more permissive on these kind of line breaks, IMO it helps me a lot to improve readability and reduce confusions from any logic's execution points.

**Comment by neongreen** [comment]
<img width="747" alt="image" src="https://user-images.githubusercontent.com/1523306/228360803-1d2efd85-acda-4313-a7ca-3b70fd3b9208.png">

Would be nice for my usecase because currently VSCode-inserted type signatures make my code constantly reflow

**Comment by larry0x** [comment]
+1 on this, although I want to use this option to prevent precisely the "nicer" way you suggested (I dislike it).

**Comment by pierrelegall** [comment]
Big +1 on this issue (with or without inlay hints issue).

However, I see two different rustfmt rules discussed here:
- line break after `=` for multiline assignment
- prefer line break after `=` if enough over multiline the right side of the expression for assignment (from @seanpianka)

**Comment by minoism** [comment]
Just wanted to add another +1 for this. I often run into situations where `rustfmt` breaks a line right after the `=` on a `let` binding. It would be great to have an option to control this behavior and keep the assignment on a single line.
